libp2p/src/builder.rs

```diff
@@ -0,0 +1,1741 @@
+                <SecUpgrade::Upgrade as InboundUpgrade<Negotiated<libp2p_tcp::$path::TcpStream>>>::Future: Send,
+                <SecUpgrade::Upgrade as OutboundUpgrade<Negotiated<libp2p_tcp::$path::TcpStream>>>::Future: Send,
+                <<<SecUpgrade as IntoSecurityUpgrade<libp2p_tcp::$path::TcpStream>>::Upgrade as UpgradeInfo>::InfoIter as IntoIterator>::IntoIter: Send,
+                <<SecUpgrade as IntoSecurityUpgrade<libp2p_tcp::$path::TcpStream>>::Upgrade as UpgradeInfo>::Info: Send,
+
+                MuxStream: StreamMuxer + Send + 'static,
+                MuxStream::Substream: Send + 'static,
+                MuxStream::Error: Send + Sync + 'static,
+                MuxUpgrade: IntoMultiplexerUpgrade<SecStream>,
+                MuxUpgrade::Upgrade: InboundUpgrade<Negotiated<SecStream>, Output = MuxStream, Error = MuxError> + OutboundUpgrade<Negotiated<SecStream>, Output = MuxStream, Error = MuxError> + Clone + Send + 'static,
+                <MuxUpgrade::Upgrade as InboundUpgrade<Negotiated<SecStream>>>::Future: Send,
+                <MuxUpgrade::Upgrade as OutboundUpgrade<Negotiated<SecStream>>>::Future: Send,
+                MuxError: std::error::Error + Send + Sync + 'static,
+                <<<MuxUpgrade as IntoMultiplexerUpgrade<SecStream>>::Upgrade as UpgradeInfo>::InfoIter as IntoIterator>::IntoIter: Send,
+                <<MuxUpgrade as IntoMultiplexerUpgrade<SecStream>>::Upgrade as UpgradeInfo>::Info: Send,
+            {
+                Ok(SwarmBuilder {
+                    phase: QuicPhase {
+                        transport: libp2p_tcp::$path::Transport::new(tcp_config)
+                            .upgrade(libp2p_core::upgrade::Version::V1Lazy)
+                            .authenticate(
+                                security_upgrade
+                                    .into_security_upgrade(&self.keypair)
+                                    .map_err(Box::new)?,
+                            )
+                            .multiplex(multiplexer_upgrade.into_multiplexer_upgrade())
+                            .map(|(p, c), _| (p, StreamMuxerBox::new(c))),
+                    },
+                    keypair: self.keypair,
+                    phantom: PhantomData,
+                })
+            }
+        }
+    };
+}
+
+impl_tcp_builder!("async-std", AsyncStd, async_io);
+impl_tcp_builder!("tokio", Tokio, tokio);
+
+pub trait IntoSecurityUpgrade<C> {
+    type Upgrade;
+    type Error: std::error::Error + 'static;
+
+    fn into_security_upgrade(self, keypair: &Keypair) -> Result<Self::Upgrade, Self::Error>;
+}
+
+impl<C, T, F, E> IntoSecurityUpgrade<C> for (F,)
+where
+    F: for<'a> FnOnce(&'a Keypair) -> Result<T, E>,
+    E: std::error::Error + 'static,
+{
+    type Upgrade = T;
+    type Error = E;
+
+    fn into_security_upgrade(self, keypair: &Keypair) -> Result<Self::Upgrade, Self::Error> {
+        self.0(keypair)
+    }
+}
```

Comment:
Note that one can not implement `IntoSecurityUpgrade` for `F` directly, as such implementation would conflict with the implementation on `(F1, F2)`.